Use FullscreenAd API

ricealexanderb opened the pull request:
One way to add an example of the fullscreen ad API would have been to create another entry on the main screen of this app, after Interstitial and Rewarded. But on Android they have a toggle switch, so instead I added support for FullscreenAd on the the app's existing AdController. There are two ad variables, the load and show functions implement both versions, and the controller supports the delegate protocols for both the old and new API. 

I've made these changes just to the Interstitial Ad in the SwiftUI version of the demo app. Instead of applying every suggestion across four files I want to get this first one right and then I'll model the other three files (SwiftUI Rewarded, Swift Interstitial, Swift Rewarded) off of this one.

The difference between these two screenshots is that I used `.padding(.horizontal)` in the second one. I guess it looks ok like this... I wanted to at least see how it would look if the text label were next to the toggle switch like it is in the Android app, but I couldn't figure out how to do that in SwiftUI.
I'm open to other layout suggestions if you have ideas.

![without-padding](https://github.com/ChartBoost/chartboost-mediation-ios-sdk-demo/assets/8443666/5b7c225d-6995-4b4b-87b2-49f46e27a6b0)
![with-padding](https://github.com/ChartBoost/chartboost-mediation-ios-sdk-demo/assets/8443666/45220381-f843-4d76-b8c9-e83238c05e9f)

ricealexanderb commented on swiftui/ChartboostMediationDemo/Ads/Interstitial/InterstitialAdController.swift:
```diff
@@ -44,47 +50,105 @@ class InterstitialAdController: NSObject, ObservableObject {
+                    self.log(action: "Load error", placementName: placementName, error: result.error)
+
+                    // `.failed` requires a non-optional error type
+                    let error = result.error ?? NSError(domain: "com.chartboost.mediation.demo", code: 0)
```
I'd like to use something that makes more sense than `code: 0` here but I didn't see any NSError Codes that seemed applicable. Does anyone know a generic NSError Code that could work here?  All the 'unknown' errors I saw were domain-specific like `NSFileReadUnknownError`

daniel-barros replied:
Why don't you make the error parameter optional in ActivityState.failed ? It will be cleaner on the point of use.

ricealexanderb commented on swiftui/ChartboostMediationDemo/Ads/Interstitial/InterstitialAdController.swift:
```diff
@@ -44,47 +50,105 @@ class InterstitialAdController: NSObject, ObservableObject {
+            let request = ChartboostMediationAdLoadRequest(placement: placementName, keywords: keywords)
+            // Load the fullscreen ad, which will make a request to the network. Upon completion, a
+            // ChartboostMediationFullscreenAdLoadResult will be passed to the completion block.
+            chartboostMediation.loadFullscreenAd(with: request) { [weak self] result in
```
This `[weak self]` is only here because I saw [that's how it's done in Canary](https://github.com/ChartBoost/ios-helium-sdk/blob/fa270e05fb4fe8ac33b1c183ee6d98e3cf8158f0/HeliumCanary/HeliumCanary/Placements/Types/Fullscreen/FullscreenPlacementViewController.swift#L158).

In the case of `ChartboostMediationFullscreenAd.show(...)` I think I see why it's necessary: because `self` holds a reference to the ad, which holds a reference to the closure, so the closure holding a strong reference to `self` would cause a cycle.

But `chartboostMediation.loadFullscreenAd(...)` is a call to the SDK, so I don't think that leads to a cycle. Is `weak self` necessary because `chartboostMediation` will outlive this ad controller and it might simply hold onto that reference and prevent this object from ever deallocating?

daniel-barros replied:
Your last statement is correct and why the `[weak self]` is necessary here. Otherwise the controller will be kept artificially alive until the load finishes, which may be long after the user has moved on from this screen.

daniel-barros commented on swiftui/ChartboostMediationDemo/Ads/Interstitial/InterstitialAdController.swift:
```diff
@@ -44,47 +50,105 @@ class InterstitialAdController: NSObject, ObservableObject {
+                guard let self = self else { return }
+                if let ad = result.ad {
+                    ad.delegate = self
+                    self.fullscreenAd = result.ad
```
```suggestion
                    self.fullscreenAd = ad
```

daniel-barros commented on swiftui/ChartboostMediationDemo/Ads/Interstitial/InterstitialAdController.swift:
```diff
@@ -44,47 +50,105 @@ class InterstitialAdController: NSObject, ObservableObject {
+                    // Notify the demo UI
+                    activityState = .idle
+                } else {
+                    fullscreenAd = nil
```
At the beginning of this method there is a safeguard to prevent multiple taps on the load button from triggering multiple loads if a load is already ongoing.
See the line `guard interstitialAd == nil else...`
This doesn't work anymore for the fullscreen APIs. I suggest we create a flag (e.g. `private var isLoading = false`) and use it instead.

ricealexanderb replied:
I think I see what you're saying: when I was managing both the `interstitialAd` and `fullscreenAd` in the same controller, I didn't extend the safeguard used here to also cover `fullscreenAd`. So if I were keeping those two different ad types in the same class, it would make sense to use a `isLoading` that worked for both of them.

But since `fullscreenAd` is split off into its own file now, I think it's ok to just keep the original logic?

daniel-barros replied:
Definitely. Just make sure the new file has some similar logic.

daniel-barros commented on swiftui/ChartboostMediationDemo/Ads/Interstitial/InterstitialAdController.swift:
```diff
@@ -44,47 +50,105 @@ class InterstitialAdController: NSObject, ObservableObject {
+
+            // Show the ad using the specified view controller.  Upon completion, instead of using a delegate
+            // method a ChartboostMediationAdShowResult will be passed to the completion block.
+            self.fullscreenAd?.show(with: viewController, completion: { [weak self] result in
```
```suggestion
            fullscreenAd.show(with: viewController, completion: { [weak self] result in
```

daniel-barros reviewed (commented):
What about separating the two implementations, creating a new FullscreenAdController.swift file?
This will be clearer for people reading this. A publisher will either integrate the old or the new APIs, not both, and our purpose here is to showcase how to integrate our SDK.
Plus, when 5.0 comes, we just have to remove the old file.

daniel-barros commented:
Also the second screenshot looks good to me. If possible, can you apply the brand green color as the tint color of the toggle for consistency?

daniel-barros commented:
> When separating the two implementations so that `FullscreenAdController` would be a separate class, I looked for a way to make the toggle switch on the individual placement screens switch between a `FullscreenAdView` and a `Interstitial/RewardedAdView`. I think the way to do that would be to put the toggle switch on a wrapper view that swaps out a subview that is either a `FullscreenAdView` or a `Interstitial/RewardedAdView` Maybe this or something similar is still the best choice, I don't know.
> 
> But then I thought it might make more sense to move the switch to the main screen like this
> 
> ![Screenshot 2023-09-21 at 10 51 13 AM](https://user-images.githubusercontent.com/8443666/269710379-62dba941-86f5-4e2a-a11e-38607f47a765.png)
> 
> And there are any number of ways to make the list of ad types change when the toggle changes, but the thing I'm stuck on is how to do it without making big changes to the existing code.
> 
> My best guess so far would be to nest two different enums inside AdType like this:
> 
> ```swift
> enum AdType: CaseIterable {
>     enum OldAPI: CaseIterable {
>         case banner
>         case interstitial
>         case rewarded
>     }
>     enum NewAPI: CaseIterable {
>         case banner
>         case fullscreenInterstitial
>         case fullscreenRewarded
>     }
> }
> ```
> 
> ...and iterate over `AdType.OldAPI.allCases` or `AdType.NewAPI.allCases` depending on the toggle setting. Then [the switch statement in AdTypeSelectionView](https://github.com/ChartBoost/chartboost-mediation-ios-sdk-demo/blob/1170e2c29d33c65f92265c5750b00ada3384e66c/swiftui/ChartboostMediationDemo/Ads/AdTypeSelectionView.swift#L56) would have six cases, three for each API subgroup.
> 
> That breaks some of the String and rawValue stuff, but I assume I'd find workarounds. My main aesthetic problem with this is that it includes `banner` in both lists even though the banner API hasn't changed.
> 
> I'm looking for guidance here because I suspect there are better options I'm not seeing. Swapping out a subview in SwiftUI might be a better choice than I realize, or there might be more elegant ways of handling two related sets of items in a single enum type, or there might be a way to refactor AdTypeSelectionView that doesn't seem too drastic. (I'm not opposed to drastic, I just don't want to re-write a bunch of it if there's a much better solution that someone will point out as soon as I push my changes)

Without looking too deep into this, my recommendation is:
- Keep the existing ad type enum as it.
- Do not duplicate the views. Keep the current InterstitialAdView and rewarded, but have the controller be injected on init.
- Wherever we create the views depending on the ad type, pass either the old or the new controller depending on the value of a boolean flag that represents the "use fullscreen api" toggle.

@ricealexanderb let me know if any of this doesn't make sense to you or if I have missed some nuances which I may have.

ricealexanderb commented:
I initially abandoned that approach because the AdViews are initialized in [this extension to AdType](https://github.com/ChartBoost/chartboost-mediation-ios-sdk-demo/blob/1170e2c29d33c65f92265c5750b00ada3384e66c/swiftui/ChartboostMediationDemo/Ads/AdTypeSelectionView.swift#L51), and I didn't see a way to access the toggle value from there.

daniel-barros commented:
> I initially abandoned that approach because the AdViews are initialized in [this extension to AdType](https://github.com/ChartBoost/chartboost-mediation-ios-sdk-demo/blob/1170e2c29d33c65f92265c5750b00ada3384e66c/swiftui/ChartboostMediationDemo/Ads/AdTypeSelectionView.swift#L51), and I didn't see a way to access the toggle value from there.

I would move that into a function inside AdTypeSelectionView.
Having a model like AdType know how to create views is not great conceptually anyway

ricealexanderb commented:
I've moved FullscreenAd into its own controller and view, and I've refactored AdType.
If this looks good to you, I'll make the same changes to the storyboard version of the app.

daniel-barros commented on swiftui/ChartboostMediationDemo/Ads/AdTypeSelectionView.swift:
```diff
@@ -28,10 +29,14 @@ struct AdTypeSelectionView: View {
                         Spacer()
                     }
-
+                    HStack {
+                        Toggle("Use Fullscreen API", isOn: $useFullscreenApi)
+                            .padding(.horizontal)
+                            .tint(.accentColor)
+                    }
```
nit: is it necessary to embed the Toggle inside a HStack? It's only one view so it shouldn't make a difference

daniel-barros commented on swiftui/ChartboostMediationDemo/Ads/AdTypeSelectionView.swift:
```diff
@@ -46,27 +51,30 @@ struct AdTypeSelectionView: View {
-                        InterstitialAdView()
-                    case .rewarded:
-                        RewardedAdView()
+    func adView(forAdType adType: AdType) -> some View {
```
nit: you could use @ViewBuilder to avoid having to explicitly declare a `adView` variable and return it. This makes the function work like other SwiftUI properties (e.g. body) and use the SwiftUI DSL instead of vanilla Swift

ricealexanderb replied:
Thanks, I have a lot to learn about SwiftUI and I appreciate the tip.

daniel-barros commented on swiftui/ChartboostMediationDemo/Ads/Banner/BannerAdView.swift:
```diff
@@ -86,10 +86,14 @@ struct BannerAdView: View {
     }
+
+    init(placementName: String) {
+        self._controller = StateObject(wrappedValue: BannerAdController(placementName: placementName))
```
This isn't a great practice in SwiftUI. A state object is maintained as long as the associated view is alive, but the BannerAdView struct itself might be recreated several times.
Can you instead pass the controller as a dependency?
You can achieve this by removing this custom init and replacing the controller definition by:
```
@ObservedObject let controller: BannerAdController
```

daniel-barros replied:
Actually nvm, I thought you were going to reuse the same view types and pass different controllers.
But I see you've created a new FullscreenAdView instead.
If you want to go that route you can then leave things the way they were before and have the views instantiate their controllers directly on their definition.

ricealexanderb replied:
Ok, I probably will leave that alone then, but I think I see what you were saying in the first comment.
A design where a single "AdView" type accepted different controller types would have been a good way to structure the app.

daniel-barros commented on swiftui/ChartboostMediationDemo/Ads/Fullscreen/FullscreenAdView.swift:
```diff
@@ -0,0 +1,89 @@
+            Image("Interstitial")
+                .padding(.vertical, 32)
+
+            Text("A full screen interstitial advertisement must first be loaded.")
```
should we add another dependency to this view named "format" or something like that, and use it to differentiate between "interstitial" and "rewarded"?
This new view is supposed to serve for both formats but now it seems to be hardcoded to display interstitial info only

ricealexanderb commented on swift/ChartboostMediationDemo/Ads/AdTypeSelectionViewController.swift:
```diff
@@ -65,14 +68,51 @@ class AdTypeSelectionViewController: UIViewController {
+        guard let viewController = adView(forAdType: adType) else {
+            // Alert the user that we can't proceed
+            let alert = UIAlertController(title: "Error",
+                                          message: "Could not create view for ",
```
Incomplete message

ricealexanderb replied:
Could also do assertion instead of an alert popup

ricealexanderb commented on swift/ChartboostMediationDemo/Ads/Fullscreen/FullscreenAdController.swift:
```diff
@@ -0,0 +1,171 @@
+            placementName = "AllNetworkInterstitial"
+        case .rewarded:
+            placementName = "AllNetworkRewarded"
+        // If adType is nill or not a fullscreen type, give an empty placement name that won't load
```
nill -> nil

ricealexanderb commented on swift/ChartboostMediationDemo/Ads/AdTypeSelectionViewController.swift:
```diff
@@ -34,11 +34,12 @@ class AdTypeSelectionViewController: UIViewController {
-        case interstitial
-        case rewarded
+    /// Current state of the "Use Fullscreen API" switch
+    var useFullscreenAPI: Bool = false
```
Is this separate variable necessary?

ricealexanderb commented on swift/ChartboostMediationDemo/Ads/Fullscreen/FullscreenAdController.swift:
```diff
@@ -0,0 +1,171 @@
+        activityDelegate?.activityDidStart()
+
+        // loadFullscreenAd expects keywords to be `[String : String]` instead of `HeliumKeywords?`.
+        let keywords = keywords?.dictionary ?? [:]
```
Instead of having keywords = nil passed in, just build an empty dictionary to pass and leave a comment about how this is where you'd add any keywords

kendallrogers commented on swiftui/ChartboostMediationDemo/Ads/Fullscreen/FullscreenAdController.swift:
```diff
@@ -0,0 +1,156 @@
+                self.log(action: "Load error", placementName: placementName, error: result.error)
+
+                // `.failed` requires a non-optional error type
+                let error = result.error ?? NSError(domain: "com.chartboost.mediation.demo", code: 0)
```
```suggestion
                let error = result.error
```

If the error object in the `.failed` state is optional, we don't need the default error here.

kendallrogers commented on swiftui/ChartboostMediationDemo/Ads/Banner/BannerAdView.swift:
```diff
@@ -86,10 +86,14 @@ struct BannerAdView: View {
             case .failed(let message, let error):
                 isBusy = false
-                failureMessage = "\(message): \(error.localizedDescription)"
+                failureMessage = "\(message): \(error?.localizedDescription ?? "")"
```
```suggestion
if let error {
                failureMessage = "\(message): \(error.localizedDescription)"
} else {
                failureMessage = message
}
```

You might also be able to handle this with pattern matching in the switch statement.
